addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +343,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
+			// util.GeneratePackageInstallName() follows the same pattern.
+			Name: util.GeneratePackageInstallName(cluster.Name, packageRefName),
+			// kapp-controller PackageInstall CR is installed under the same namespace as tanzuClusterBootstrap. The namespace
+			// is also the same as where the cluster belongs.
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: clusterapiv1beta1.GroupVersion.String(),
+					Kind:       cluster.Kind,
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
 	}
 
-	if pkg.ValuesFrom.ProviderRef != nil {
-		gvr, err := r.getGVR(schema.GroupKind{Group: *pkg.ValuesFrom.ProviderRef.APIGroup, Kind: pkg.ValuesFrom.ProviderRef.Kind})
+	pkgiMutateFn := func() error {
+		// TODO: Followup on the following fields, only add them if needed.
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1677
+		// if ipkg.ObjectMeta.Annotations == nil {
+		//	 ipkg.ObjectMeta.Annotations = make(map[string]string)
+		// }
+		// ipkg.ObjectMeta.Annotations[addontypes.YttMarkerAnnotation] = ""
+		pkgi.Spec.SyncPeriod = &metav1.Duration{Duration: r.Config.PkgiSyncPeriod}
+		pkgi.Spec.PackageRef = &kapppkgiv1alpha1.PackageRef{
+			// clusterBootstrap.Spec.Kapp.RefName is Package.Name. I.e., kapp-controller.tanzu.vmware.com.0.28.0+vmware.1-tkg.1-rc.1
+			// PackageInstall.Spec.PackageRef looks for the Package.Spec.refName which is a short name of the full Package.Name
+			// packageRefName and packageVersion are fetched from the Package CR.
+			RefName: packageRefName,
+			VersionSelection: &versions.VersionSelectionSemver{
+				Constraints: packageVersion,
+				Prereleases: &versions.VersionSelectionSemverPrereleases{},
+			},
+		}
+		// Adding the cluster reference to PackageInstall spec to instruct kapp-controller where to deploy
+		// the underlying resources
+		clusterKubeconfigDetails := util.GetClusterKubeconfigSecretDetails(cluster)
+		pkgi.Spec.Cluster = &kappctrlv1alpha1.AppCluster{
+			KubeconfigSecretRef: &kappctrlv1alpha1.AppClusterKubeconfigSecretRef{
+				Name: clusterKubeconfigDetails.Name,
+				Key:  clusterKubeconfigDetails.Key,
+			},
+		}
+		secretName, err := r.GetDataValueSecretNameFromBootstrapPackage(clusterBootstrap.Spec.Kapp, cluster.Namespace)
 		if err != nil {
-			log.Error(err, "failed to getGVR")
-			return nil, err
+			return err
+		}
+		pkgi.Spec.Values = []kapppkgiv1alpha1.PackageInstallValues{
+			{SecretRef: &kapppkgiv1alpha1.PackageInstallValuesSecretRef{
+				Name: secretName},
+			},
 		}
-		provider, err := r.dynamicClient.Resource(*gvr).Namespace(namespace).Get(r.context, pkg.ValuesFrom.ProviderRef.Name, metav1.GetOptions{}, "status")
+		return nil
+	}
 
-		if err != nil {
-			log.Error(err, "unable to fetch provider", "provider", pkg.ValuesFrom.ProviderRef, "gvr", gvr)
-			return nil, err
+	_, err = controllerutil.CreateOrPatch(r.context, r.Client, pkgi, pkgiMutateFn)
```

Comment:
Question: As the installation is local in the management cluster (under workload cluster namespace), we use ```r.client``` as opposed to ```remoteClient```. Right?

Reply:
yes, that is correct.